hazelcast/src/test/java/com/hazelcast/instance/DefaultNodeExtensionTest.java

```diff
@@ -0,0 +1,152 @@
+        Version majorPlusOne = new Version(currentVersion.getMajor() + 1, currentVersion.getMinor(), currentVersion.getPatch());
+        assertFalse(nodeExtension.isNodeVersionCompatibleWith(majorPlusOne));
+    }
+
+    @Test
+    public void test_joinRequestAllowed_whenSameVersion()
+            throws UnknownHostException {
+        JoinRequest joinRequest = new JoinRequest(Packet.VERSION, BUILD_INFO.getBuildNumber(), node.getVersion(),
+                new Address("127.0.0.1", 9999), UuidUtil.newUnsecureUuidString(), false, null, null, null, null);
+        nodeExtension.validateJoinRequest(joinRequest);
+    }
+
+    @Test
+    public void test_joinRequestAllowed_whenOtherPatchVersion()
+            throws UnknownHostException {
+        Version otherPatchVersion = Version
+                .of(node.getVersion().getMajor(), node.getVersion().getMinor(), node.getVersion().getPatch() + 1);
+        JoinRequest joinRequest = new JoinRequest(Packet.VERSION, BUILD_INFO.getBuildNumber(), otherPatchVersion,
+                new Address("127.0.0.1", 9999), UuidUtil.newUnsecureUuidString(), false, null, null, null, null);
+        nodeExtension.validateJoinRequest(joinRequest);
+    }
+
+    @Test
+    public void test_joinRequestFails_whenOtherMinorVersion()
+            throws UnknownHostException {
+        Version otherPatchVersion = Version
+                .of(node.getVersion().getMajor(), node.getVersion().getMinor() + 1, node.getVersion().getPatch());
+        JoinRequest joinRequest = new JoinRequest(Packet.VERSION, BUILD_INFO.getBuildNumber(), otherPatchVersion,
+                new Address("127.0.0.1", 9999), UuidUtil.newUnsecureUuidString(), false, null, null, null, null);
+        expected.expect(VersionMismatchException.class);
+        nodeExtension.validateJoinRequest(joinRequest);
+    }
+
+    @Test
+    public void test_joinRequestFails_whenOtherMajorVersion()
+            throws UnknownHostException {
+        Version otherPatchVersion = Version
+                .of(node.getVersion().getMajor() + 1, node.getVersion().getMinor(), node.getVersion().getPatch());
+        JoinRequest joinRequest = new JoinRequest(Packet.VERSION, BUILD_INFO.getBuildNumber(), otherPatchVersion,
+                new Address("127.0.0.1", 9999), UuidUtil.newUnsecureUuidString(), false, null, null, null, null);
+        expected.expect(VersionMismatchException.class);
+        nodeExtension.validateJoinRequest(joinRequest);
+    }
+
+    @Test
+    public void test_clusterVersionListener_invokedOnRegistration()
+            throws UnknownHostException {
+        final CountDownLatch latch = new CountDownLatch(1);
+        ClusterVersionListener listener = new ClusterVersionListener() {
+            @Override
+            public void onClusterVersionChange(Version newVersion) {
+                latch.countDown();
+            }
+        };
+        assertTrue(nodeExtension.registerListener(listener));
+        assertOpenEventually(latch);
+    }
+
```

Comment:
Could we have a test to guard the the listener will get node's codebase version if cluster.getVersion is null?
Otherwise the versioned serialization will be screwed.